webapps/frontend/app/widgets/searchbar/viewmodel.js

```diff
@@ -0,0 +1,211 @@
+                    // Chain these with OR
+                    var orItems = [];
+                    var orQuery = {
+                        'type': 'OR',
+                        'items': orItems
+                    };
+                    $.each(operator, function(opIndex, op){
+                        orItems.push([field, op, value])
+                    });
+                    queryItems.push(orQuery)
+                } else {
+                    queryItems.push([field, operator, value])
+                }
+            });
+            return query
+        };
+
+        var extractField = function(field, fieldMap) {
+            if (fieldMap === undefined) { fieldMap = self.fieldMap}
+            if (field in fieldMap) {return fieldMap[field]}
+            return field
+        };
+
+        var testQueryBuilder = function() {
+            var defaultField = 'guid';
+            var fieldMap = {};
+            var tests = [['', []],
+                ['test', [['guid', 'EQUALS', 'test']]],
+                ['test:IN"test"', [['test', 'IN', 'test']]],
+                ['"test":!=test', [['test', 'NOT_EQUALS', 'test']]],
+                ['test', [['test', 'EQUALS', 'test'], ['guid', 'EQUALS', 'test']], {items: [['test', 'EQUALS', 'test']]}]
+            ];
+            $.each(tests, function(index, testCase){
+                var defaultQuery;
+                if (testCase.length === 3) {
+                    defaultQuery = testCase[2]
+                }
+                var query = buildQuery(testCase[0], defaultField, fieldMap, defaultQuery);
+                if (generic.objectEquals(query.items, testCase[1]) === false) {
+                    throw new Error(query.items.toString() + testCase[1].toString())
+                }
+            })
+        };
+
+        var cleanQuotes = function(text) {
+            // Slicing is faster than regex :)
+            if (text.startsWith("'") || text.startsWith('"')) {
+                text = text.slice(1)
+            }
+            if (text.endsWith("'") || text.endsWith('"')) {
+                text = text.slice(0, -1)
+            }
+            return text
+        };
+
+        // Durandal
+        /**
+         * Activator for the widget
+         * @param settings: settings to supply
+         * Settings is an object which should have:
+         * - return: the observable to use to return the searching api data
+         * And optionally has:
+         * - placeholder: placeholder to show on the input field
+         * - fieldMap: object which has the options as keys and the field as value (eg. {sr_guid: storagerouter_guid}
+         * - default field: default field to map to when no fields are specified. eg( my_disk sr_guid:a_guid) with default field = name would result into: {name: my_disk sr_guid:a_guid}
+         * @type settings: object
+         */
+        self.activate = function(settings) {
+            if (!settings.hasOwnProperty('query')) {
```

Comment:
Also verify whether `defaultQuery` has been provided since you require this 3 lines down

Reply:
No need, it will result in undefined when it does not exist (a bit nasty though)